File: src/Software Engineer Prepare Kit/21/solution.go
package main

func CountIsolatedCommunicationGroups(links [][]int32, n int32) int32 {
	components := n

	parent := make([]int32, n)
	for i := range n {
		parent[i] = i
	}

	var find func(i int32) int32
	find = func(i int32) int32 {
		if parent[i] == i {
			return i
		}

		parent[i] = find(parent[i])

		return parent[i]
	}

	for _, link := range links {
		rootA := find(link[0])
		rootB := find(link[1])

		if rootA != rootB {
			parent[rootB] = rootA
			components--
		}
	}

	return components
}

File: src/Software Engineer Prepare Kit/21/solution.ts
export default function countIsolatedCommunicationGroups(links: number[][], n: number): number {
	let components: number = n

	const parent: number[] = Array.from({ length: n }, (_, i) => i)

	const find = (i: number): number => {
		if (parent[i] === i) {
			return i
		}

		parent[i] = find(parent[i]!)

		return parent[i]
	}

	for (const link of links) {
		const rootA = find(link[0]!)
		const rootB = find(link[1]!)

		if (rootA !== rootB) {
			parent[rootB] = rootA
			components--
		}
	}

	return components
}
